src/ast/passes/codegen_llvm.cpp

```diff
@@ -2283,29 +2310,32 @@ void CodegenLLVM::visit(AssignMapStatement &assignment)
   Value *val, *expr;
   expr = expr_;
   auto [key, scoped_key_deleter] = getMapKey(map);
-  if (shouldBeInBpfMemoryAlready(assignment.expr->type)) {
-    if ((assignment.expr->type.IsStringTy() ||
-         assignment.expr->type.IsTupleTy()) &&
-        assignment.expr->type.GetSize() != map.type.GetSize()) {
+  auto &expr_type = assignment.expr->type;
+  if (shouldBeInBpfMemoryAlready(expr_type)) {
+    if (expr_type.GetSize() != map.type.GetSize()) {
       val = b_.CreateAllocaBPF(map.type, map.ident + "_val");
       b_.CreateMemsetBPF(val, b_.getInt8(0), map.type.GetSize());
-      b_.CreateMemcpyBPF(val, expr, assignment.expr->type.GetSize());
-      self_alloca = true;
-    } else
+      if (expr_type.IsTupleTy()) {
+        createTupleCopy(expr_type, map.type, val, expr);
+        self_alloca = true;
+      } else if (expr_type.IsStringTy()) {
+        b_.CreateMemcpyBPF(val, expr, expr_type.GetSize());
+        self_alloca = true;
+      } else {
+        val = expr;
+      }
```

Comment:
I know the logic is equivalent, but the new layout highlights a potential issue more clearly: it wasn't expected that any non-tuple, non-string type could have a mismatched size. i.e. this `else` branch probably won't work right.

I'm thinking that a `buf` could break it, but should we just turn it into a `LOG(BUG)` for now?

Reply:
I wonder if it would make it through the semantic_analyser pass. I'm fine to turn it into a `LOG(BUG)` or a `LOG(WARNING)` if we're worried it might break existing scripts.